### PR TITLE
Distillation

### PR DESCRIPTION
**Pull request by mitchellnw**
Still need to test, made changes again but after CoCa merge

### CODE REVIEW

**Comment by mitchellnw** [comment]
this is with equal weight on both losses which is also what coca does so would be good if this gives high accuracy and we can stick with it

https://wandb.ai/dogml/open_clip6/runs/distill_l14_32b_laion2b_v4

**Comment by mitchellnw** [comment]
got 67.64% when distilling L/14 into B/32 with 13b samples seen so seems good to go (+ > 4%). a few things are not supported and have fixmes such as
- distillation + grad accum
- distillation + coca
but these should probably be follow up PRs.

**Comment by mitchellnw** [comment]
anything required before merging? understand if no plans to merge due to too many features

**Comment by rsomani95** [comment]
@mitchellnw the ImageNet top-1 reported was at the end of the full training run. I'm curious if you measured that every few epochs as well? I'm wondering what's the accuracy growth curve looks like - trying to build intuition for how many image samples one would need to be able to distill. 

On a related note, have you done any experiments distilling either of these scenarios:
1. A `timm` image encoder model that's been pretrained on ImageNet
2. (1) initialised with the _same_ text encoder as the teacher model

My intuition is that both of these scenarios would lead to faster convergence. In my use case, we need to run the image encoder much more often than the text encoder, so it's okay to have a heavier text model as described in (2)

**Comment by mitchellnw** [comment]
Sorry we never ran more experiments on this, we only have the final number as reported! I would be very curious to see what happens.

**Comment by rsomani95** [comment]
@mitchellnw got it, thanks for the response. Do you think those experiments make sense? I'd love your perspective on it and if you have any suggestions / gotchas to look out for. I will be running some form of these experiments, but have limited compute (3x 3090s), so any time saved with better assumptions would be very welcome :)

Specifically, I'll be training an ImageNet pretrained `ghostnet_100` variant with a pretrained `convnext_base_w` text encoder, and distilling from `convnext_base_w` for starters.

**Comment by mitchellnw** [comment]
The distillation code we have right now distills in the contrastive image-text setting so I don't believe an imagenet pre-trained model will work

**Comment by mitchellnw** [comment]
Unless you mean to go the other way?

**Comment by rsomani95** [comment]
> The distillation code we have right now distills in the contrastive image-text setting so I don't believe an imagenet pre-trained model will work

Interesting. Do you think a randomly initialised img encoder is better to start with than one that's been pre-trained on ImageNet? Or maybe an img encoder that's been fine-tuned to be a bit more aligned with a CLIP image encoder?

More formally, I'm trying to evaluate these three strategies for the best initialisation step for the image encoder:
1. Random initialisation
2. ImageNet pretraining
3. An ImageNet model that's been distilled from the CLIP teacher model's image encoder _only_
    - Here, the idea is to take a few million (billion?) images, compute embeddings for the teacher and student image encoder, and do a simple mean squared error loss
    - If this helps, it can be a lot cheaper and faster to do as a step before doing the contrastive image-text distillation

**Comment by mitchellnw** [comment]
If you're distilling into an encoder than I think starting that encoder with imagenet pre-trained will be better. But I also think that's not specific to distillation -- if you start CLIP training from a pre-trained encoder I also think that will be better than from scratch. I really don't know, though.

**Comment by rsomani95** [comment]
Got it. Thanks, that's helpful. Will post any notable findings here as I go along.

**Comment by abhishek0093** [comment]
Hi @mitchellnw , 
Thanks for this great addition to the repo. I was just wondering if these distilled models which you trained above are available somewhere? 
And do you think it would work if I finetune a B/16 and simulatenously distill it from L/14 on my custom dataset of ~2.5M datapoints.
